stdlib/src/handlers/aead_decrypt.rs

```diff
@@ -0,0 +1,140 @@
+
+    // Read ciphertext from memory: (num_blocks + 1) * 8 elements (data + padding)
+    let num_ciphertext_elements = (num_blocks + 1) * 8;
+    let ciphertext = read_memory_region(process, src_ptr, num_ciphertext_elements).ok_or(
+        AeadDecryptError::MemoryReadFailed {
+            addr: src_ptr,
+            len: num_ciphertext_elements,
+        },
+    )?;
+
+    // Read authentication tag: 4 elements (1 word) immediately after ciphertext
+    let tag_ptr = src_ptr + num_ciphertext_elements;
+    let tag_addr: u32 = tag_ptr
+        .try_into()
+        .ok()
+        .ok_or(AeadDecryptError::MemoryReadFailed { addr: tag_ptr, len: 4 })?;
+
+    let ctx = process.ctx();
+    let tag_word = process
+        .get_mem_word(ctx, tag_addr)
+        .map_err(|_| AeadDecryptError::MemoryReadFailed { addr: tag_ptr, len: 4 })?
+        .ok_or(AeadDecryptError::MemoryReadFailed { addr: tag_ptr, len: 4 })?;
+
+    let tag_elements: [miden_core::Felt; 4] = tag_word.into();
+
+    // Convert to reference implementation types
+    let secret_key = SecretKey::from_elements(key_word.into());
+    let nonce = Nonce::from(nonce_word);
+    let auth_tag = AuthTag::new(tag_elements);
+
+    // Construct EncryptedData
+    let encrypted_data =
+        EncryptedData::from_parts(DataType::Elements, ciphertext, auth_tag, nonce.clone());
+
+    // Decrypt using the standard reference implementation
+    // This performs tag verification internally
+    let plaintext_with_padding = secret_key.decrypt_elements(&encrypted_data)?;
+
+    // Extract only the data blocks (without padding) to push onto advice stack
+    // The MASM encrypt procedure will add padding automatically during re-encryption
+    let data_blocks_count = (num_blocks * 8) as usize;
+    let mut plaintext_data = plaintext_with_padding;
+    plaintext_data.truncate(data_blocks_count);
+
+    // Push plaintext data (WITHOUT padding) onto advice stack
+    // The padding will be added by the MASM encrypt procedure during re-encryption
+    // Note: We push in reverse order so adv_pipe reads them in forward order
+    plaintext_data.reverse();
+    let advice_stack_mutation = AdviceMutation::extend_stack(plaintext_data);
+
+    Ok(vec![advice_stack_mutation])
+}
+
+// ERROR HANDLING
+// ================================================================================================
+
+/// Error types that can occur during AEAD decryption.
+#[derive(Debug, thiserror::Error)]
+enum AeadDecryptError {
+    /// Memory read failed or address overflow.
+    #[error("failed to read memory region at addr={addr}, len={len}")]
+    MemoryReadFailed { addr: u64, len: u64 },
+
+    /// Decryption failed (wraps EncryptionError from miden-crypto).
+    #[error(transparent)]
+    DecryptionFailed(#[from] EncryptionError),
+}
+
+// TESTS
+// ================================================================================================
+
+#[test]
+fn test_event_name() {
```

Comment:
We should put the test under a `#[cfg(test)] mod tests { ... }` module, so that the test functions don't get included in the main library.